### DISCUSSION TITLE
Controlling asset backfill concurrency

### DISCUSSION THREAD

**Discussion by jPinhao-Rover** [dagster-io/dagster#18224] [Q&A]
From [limiting concurrency docs](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines#limiting-concurrency-using-tags), and [asset backfill Q&A](https://github.com/dagster-io/dagster/discussions/12251) it seems we should be able to limit concurrency on asset backfills using tags.

We would like to limit the concurrency only on certain backfills (ie. explicitly tag certain assets as needing to be backfilled 1 run at a time). We were able to do this by:

1. Define a tag `serial_backfill` and set it to the job name on each job we want to limit:

```python
test_controlled_backfill = define_asset_job(
    name="test_controlled_backfill",
    selection=AssetSelection.groups("spike_partitions"),
    tags = TAGS | {
        "serial_backfill": "test_controlled_backfill"
    }
)

test_another_controlled_backfill = define_asset_job(
    name="test_another_controlled_backfill",
    selection=AssetSelection.groups("spike_partitions"),
    tags = TAGS | {
        "serial_backfill": "test_another_controlled_backfill"
    }
)
```

2. Define an instance wide concurrency limit in our `dagster.yaml` :

```
run_coordinator:
  module: dagster.core.run_coordinator
  class: QueuedRunCoordinator

  config:
    tag_concurrency_limits:
      - key: "serial_backfill"
        limit: 1
        value:
          applyLimitPerUniqueValue: true
```

This works well, allowing only 1 concurrent backfill per job, but each job able to run it's own backfill at the same time. However there's a couple of issues:
1. This limits the backfill from _any_ asset within the job (ie. I can't backfill 2 independent assets concurrently) unless they're triggered in the same run.
2. Job's [don't support backfill_policy](https://github.com/dagster-io/dagster/discussions/18158), which we actually must have in order to control how we execute larger historical models.

To resolve these, we've tried moving the tags to the asset under the `op_tags` , but this doesn't seem to be picked up by the rules in `tag_concurrency_limits` , ie. by setting:

```
@asset(
    partitions_def=PARTITION_DEFINITION,
    group_name="spike_partitions",
    compute_kind="python",
    backfill_policy=BackfillPolicy.multi_run(max_partitions_per_run=5),
    op_tags = {
        "serial_backfill": "test_standalone_asset"
    }
)
def test_standalone_asset(context):
...
```

and running a backfill either through the job UI or asset UI, doesn't limit how many runs get triggered concurrently, it seems the `op_tags` aren't actually picked up as "job tags" and therefore might not be useable to control instance wide concurrency?

How can we define backfill concurrency limits on an asset with a `backfill_policy`, such that we can both a) prevent more than 1 backfill for the same asset running concurrently, and b) control how many partitions should be materialised in each asset backfill run?

**Comment by benpankow** [comment]
Hello - I think what you may want in your use-case is the [currently-experimental cross-run asset/op concurrency limits](https://docs.dagster.io/guides/limiting-concurrency-in-data-pipelines#limiting-opasset-concurrency-across-runs). You can apply these tags to individual assets & ops, and the limits are enforced globally such that an asset can only be materialized up to `n` times simultaneously.

```python

@asset(op_tags={"dagster/concurrency_key": "my_standalone_asset"})
def my_standalone_asset():
    ...
```

**Reply by jPinhao-Rover** [reply]
I had actually missed this, thanks for pointing it out!

However, this doesn't actually seem to work. When I trigger a backfill which forces multiple runs to start, multiple of them seem to start at the same time. The "Concurrency limits" tab does eventually show that there is 1 occupied slot and 1 active run, however multiple runs have been triggered in the meantime:

![image](https://github.com/dagster-io/dagster/assets/144812520/fa5375d8-e83c-4054-ae6b-e82e04d21561)
 

Also it's worth pointing out that this experimental feature, by not using the `tag_concurrency_limits` config with `applyLimitPerUniqueValue: true` option, and instead requiring the key to be explicitly configured via UI/cli does it make it challenging to actually configure concurrency limits per individual asset/op (you'd need to run a separate process, or during server initialisation, somehow identify all potential values for this tag and set the concurrency limit for each inidividal tag)

**Reply by the4thamigo-uk** [reply]
Oh I seem to have also stumbled on this issue. Is there a solution/workaround?

**Reply by jPinhao-Rover** [reply]
Unfortunately didn't find a better one for our case than:
1. Set a limit against a tag (eg. `serial_backfill`) which needs to be manually set by the end-user when triggering a backfill job (error prone, missable) 
2. Set a limit against `dagster/backfill` as a fallback to prevent a backfill taking over the scheduling queue if the human operator forgets to set `serial_backfill`

[a comment by an automated account is not shown]

**Reply by prha** [reply]
There's not a way to set a limit via helm chart, but there's active development happening for this feature, to implement [a default limit](https://github.com/dagster-io/dagster/issues/17145).  I'll make a note to ensure this is configurable via helm when it's launched.
